### PR TITLE
Bug fix/foxx allow post body on delete request

### PR DESCRIPTION
**Pull request by dothebart**
as @adityamukho found out we removed POST body support for DELETE; since this is very usefull for i.e. bulk-delete operations and we also do this on c++ API, re-add support for it.

### CODE REVIEW

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8567/

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8823/

**Comment by dothebart** [comment]
during developing the tests several more issues came up:

- vst doesn't set content-length headers, since its part of the protocol
  - this is wrong for HEAD-requests fixed on the server side, head requests will now emmit a content-length header. 
  - all other cases are fixed on the client side in the V8ClientConnection which now syntesizes a content-length header in these cases.
- 2 places are used for the size of reply documents. To avoid copying for HEAD requests in first place, only a counter is maintained. This leads to wrong content-length values in case of HEAD requests from foxx, since they write the reply body anyways, and don't set the counter. This is now fixed in the actions interface. 

- Http2 
  - fuerte would swallow the content-length header, and not pass it along to its client
  - in case of HEAD requests Arangod would not emmit a content-length header. 

- in case of missuse of the foxx interface, non rfc-conformant http replies would be sent. This is now logged ( x7f45) and corrected
- V8ClientConnection reply body handling wasn't good for the `raw` cases. It would only create a Javascript string - which is wrong. It will now: 
  - parse velocystream replies and emmit them
  - create Buffer-Objects from binary data

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8826/

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8827/

**Comment by ObiWahn** [comment]
Has this PR tests? What does it change exactly? If I was @omar or @lservini I would not know how to verify this. Please make the QA phase shorter by providing sufficient descriptions.

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8827/

The changes in this PR were found neccessary when writing the test. So - yes this has a test.

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8828/

File: js/server/modules/@arangodb/foxx/router/validation.js [hunk 1]
```diff
@@ -56,7 +56,7 @@ exports.parseRequestBody = function parseRequestBody (def, req) {
   let body = req.body;
 
   if (!def.contentTypes) {
-    assert(!body || !body.length, 'Unexpected request body');
+     assert(!body || !body.length, 'Unexpected request body');
```
Comment (pluma) [inline js/server/modules/@arangodb/foxx/router/validation.js]: Please remove excess white space here.

Comment (pluma) [reply js/server/modules/@arangodb/foxx/router/validation.js]: lgtm otherwise

File: lib/Rest/GeneralResponse.cpp [hunk 2]
```diff
@@ -31,6 +31,77 @@
...
+    default: {
+      int group = ((int)code) / 100;
+      switch (group) {
+      case 1:
+      case 2:
+      case 3:
+      case 4:
+      case 5:
+        return true;
+        if (static_cast<int>(code) != 0) {
```
Comment (ObiWahn) [inline lib/Rest/GeneralResponse.cpp]: This looks wrong! Why is the `return true` followed by another `if` and `return`?

File: arangod/GeneralServer/H2CommTask.cpp [hunk 3]
```diff
@@ -576,10 +576,14 @@ void H2CommTask<T>::queueHttp2Responses() {
...
+      std::string len;
+      len = std::to_string(res.bodySize());
+      nva.push_back({(uint8_t*)"content-length", (uint8_t*)len.c_str(), 14,
+                     len.length(), NGHTTP2_NV_FLAG_NO_COPY_NAME});
+    } else if (res.generateBody() &&
+               ::expectResponseBody(static_cast<int>(res.responseCode()))) {
+      std::string len;
       len = std::to_string(res.bodySize());
       nva.push_back({(uint8_t*)"content-length", (uint8_t*)len.c_str(), 14,
                      len.length(), NGHTTP2_NV_FLAG_NO_COPY_NAME});
```
Comment (jsteemann) [inline arangod/GeneralServer/H2CommTask.cpp]: The code for the `if` and the `else if` cases seems to be exactly identical.
Looks like we should unify it and put it behind an `if` statement that just is
```
if (::expectResponseBody(static_cast<int>(res.responseCode()))) {
```

File: js/server/modules/@arangodb/foxx/router/validation.js (hunk 1 above)
Comment (jsteemann) [inline js/server/modules/@arangodb/foxx/router/validation.js]:
```suggestion
    assert(!body || !body.length, 'Unexpected request body');
```

File: lib/Rest/GeneralResponse.cpp (hunk 2 above)
Comment (jsteemann) [reply lib/Rest/GeneralResponse.cpp]: This seems overcomplicated.
We are first checking individual response codes and return true if the response code is one of them.
But afterwards, we allow any response code between 100 and 599 anyway. So why the complexity?
And @ObiWahn is right in that the `if` following the `return` is not very useful.

File: lib/Rest/HttpResponse.h [hunk 4]
```diff
@@ -67,6 +67,10 @@ class HttpResponse : public GeneralResponse {
   }
   size_t bodySize() const;
 
+  void sealBody() {
+    _bodySize = _body->length();
+  };
```
Comment (jsteemann) [inline lib/Rest/HttpResponse.h]:
```suggestion
  }
```

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/8833/

**Pull request merged**